IdentityCore/src/oauth2/MSIDTokenResponse.m

```diff
@@ -130,29 +116,60 @@ - (MSIDErrorCode)oauthErrorCode
     return MSIDErrorCodeForOAuthError(self.error, MSIDErrorServerOauth);
 }
 
-- (NSDictionary *)additionalServerInfo
+#pragma mark - Protected
+
+- (MSIDIdTokenClaims *)tokenClaimsFromRawIdToken:(NSString *)rawIdToken error:(NSError **)error
 {
-    NSArray *knownFields = @[MSID_OAUTH2_ERROR,
-                             MSID_OAUTH2_ERROR_DESCRIPTION,
-                             MSID_OAUTH2_ACCESS_TOKEN,
-                             MSID_OAUTH2_TOKEN_TYPE,
-                             MSID_OAUTH2_REFRESH_TOKEN,
-                             MSID_OAUTH2_SCOPE,
-                             MSID_OAUTH2_STATE,
-                             MSID_OAUTH2_ID_TOKEN,
-                             MSID_OAUTH2_EXPIRES_IN];
-    
-    NSDictionary *additionalInfo = [_json dictionaryByRemovingFields:knownFields];
-    if (additionalInfo.count > 0)
+    return [[MSIDIdTokenClaims alloc] initWithRawIdToken:rawIdToken error:error];
+}
+
+#pragma mark - MSIDJsonSerializable
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json error:(NSError **)error
+{
+    self = [super init];
+    if (self)
     {
-        return additionalInfo;
+        if (!json)
+        {
+            if (error) *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Attempt to initialize token response with nil json", nil, nil, nil, nil, nil);
+            
+            return nil;
+        }
+        
+        _accessToken = [json msidStringObjectForKey:MSID_OAUTH2_ACCESS_TOKEN];
+        _refreshToken = [json msidStringObjectForKey:MSID_OAUTH2_REFRESH_TOKEN];
+        _expiresIn = [json msidIntegerObjectForKey:MSID_OAUTH2_EXPIRES_IN];
+        _expiresOn = [json msidIntegerObjectForKey:MSID_OAUTH2_EXPIRES_ON];
+        _tokenType = [json msidStringObjectForKey:MSID_OAUTH2_TOKEN_TYPE];
+        _scope = [json msidStringObjectForKey:MSID_OAUTH2_SCOPE];
+        _state = [json msidStringObjectForKey:MSID_OAUTH2_STATE];
+        self.idToken = [json msidStringObjectForKey:MSID_OAUTH2_ID_TOKEN];
+        _error = [json msidStringObjectForKey:MSID_OAUTH2_ERROR];
+        _errorDescription = [json msidStringObjectForKey:MSID_OAUTH2_ERROR_DESCRIPTION];
+        self.additionalServerInfo = json;
```

Comment:
nit: should we call setter explicitly? Otherwise people might want to fix this code for consistency without noticing that setters of id token and additional server info are important.

Reply:
done